### PR TITLE
[WIP] Time zone functionality

### PR DESCRIPTION
**Pull request by MusikAnimal**
**Work in progress, do not merge**

Bug: [T140306](https://phabricator.wikimedia.org/T140306)

### CODE REVIEW

File: app/assets/javascripts/components/common/time_zone.jsx [hunk 1]
```diff
@@ -0,0 +1,92 @@
+import React from 'react';
+
+const DatePicker = React.createClass({
```
Comment (ragesoss) [inline app/assets/javascripts/components/common/time_zone.jsx]: TimeZone, not DatePicker

File: db/schema.rb [hunk 2]
```diff
@@ -11,304 +10,294 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 20160930190451) do
+ActiveRecord::Schema.define(version: 20161005185412) do
```
Comment (ragesoss) [inline db/schema.rb]: Be sure to sanitize the schema.rb changes to just make the minimum change, before we merge.

File: app/models/course.rb [hunk 3]
```diff
@@ -39,6 +39,7 @@
 #  syllabus_file_size    :integer
 #  syllabus_updated_at   :datetime
 #  home_wiki_id          :integer
+#  timezone              :string           default("UTC")
```
Comment (ragesoss) [inline app/models/course.rb]: let's be sure to validate this against inclusion in the set ActiveSupport knows about.

File: db/schema.rb (hunk 2 above)
Comment (MusikAnimal) [inline db/schema.rb]: Hmm that I don't know how to do. I did find it a bit odd that after my migration this file changed so much!

Comment (ragesoss) [inline db/schema.rb]: schema.rb gets recreated every time you run `rake db:migrate`... based on the state of your own database. But in tests and production, the schema is what gets loaded to control behavior.

In this case, the only things that should change are this line defining the prefix of the latest required migration...

File: db/schema.rb [hunk 4]
```diff
@@ -11,304 +10,294 @@
...
+    t.integer  "upload_count",                        default: 0
+    t.integer  "uploads_in_use_count",                default: 0
+    t.integer  "upload_usages_count",                 default: 0
+    t.string   "syllabus_file_name"
+    t.string   "syllabus_content_type"
+    t.integer  "syllabus_file_size"
     t.datetime "syllabus_updated_at"
-    t.integer  "home_wiki_id",          limit: 4
+    t.integer  "home_wiki_id"
+    t.string   "time_zone",                           default: "UTC"
```
Comment (ragesoss) [inline db/schema.rb]: And this line, the only column that has changed.

Comment (ragesoss) [inline db/schema.rb]: So you can sort it out by reverting to the original version, and then just manually updating those two lines.

File: app/views/shared/_head.html.haml [hunk 5]
```diff
@@ -15,6 +15,8 @@
   I18n.defaultLocale = "#{I18n.default_locale}";
   I18n.locale = "#{I18n.locale}";
 
+  TimeZones = #{ActiveSupport::TimeZone.all.each_with_object({}) { |v,h| h[v.name] = v.formatted_offset }.to_json};
```
Comment (ragesoss) [inline app/views/shared/_head.html.haml]: Maybe instead of doing this translation on each page load, we should set up the array in an initializer.

**Comment by MusikAnimal** [comment]
Closing in favour of #993

**Pull request closed without merging**